sources/org.osbuild.git

```diff
@@ -0,0 +1,106 @@
+#!/usr/bin/python3
+
+import json
+import os
+import sys
+import subprocess
+import tempfile
+
+
+STAGE_DESC = "Fetch the contents of a git commit"
+STAGE_INFO = """
+This source can be used to obtain the contents of specific git
+commits via their checksum. For all specified checksums only the
+individual commit is fetched. This needs support from the git
+remote, i.e. uploadpack.allowReachableSHA1InWant but be turned
+on. That feature was introduced with git 2.5.0, released in July
+2015. Currently no fallback is provided, in case that fails.
+
+"""
+STAGE_OPTS = """
+"additionalProperties": false,
+"properties": {
+  "commits": {
+    "description": "The commits to fetch",
+    "type": "object",
+    "additionalProperties": false,
+    "patternProperties": {
+      "\b[0-9a-f]{5,64}\b": {
+        "type": "object",
+        "additionalProperties": false,
+        "required": ["remote"],
+        "properties": {
+          "remote": {
+            "type": "object",
+            "additionalProperties": false,
+            "required": ["remote"],
+            "properties": {
+              "url": {
+                "type": "string",
+                "description": "URL of the remote"
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
+"""
+
+
+def git(*args, _input=None, _gitdir=None, **kwargs):
+    args = list(args) + [f'--{k}={v}' for k, v in kwargs.items()]
+    env = os.environ.copy()
+    if _gitdir:
+        env["GIT_DIR"] = _gitdir
+    subprocess.run(["git"] + args,
+                   env=env,
+                   encoding="utf-8",
+                   stdout=sys.stderr,
+                   input=_input,
+                   check=True)
+
+
+def main(options, checksums, output):
+    commits = options["commits"]
+
+    os.makedirs(output, exist_ok=True)
+
+    for commit in checksums:
+        remote = commits[commit]["remote"]
+        url = remote["url"]
+
+        # Fetch *only* the specific commit by first creating a local
+        # bare git repository and then fetching the specified commit
+        # with a `depth` of 1. NB: this needs support from the remote
+        # repository as well (`uploadpack.allowReachableSHA1InWant`),
+        # first introduced with git 2.5.0, released in July 2015.
+        with tempfile.TemporaryDirectory() as tmp:
+            repo = f"{tmp}/repo"
+
+            git("init", "--quiet", "--bare", repo)
+
+            git("remote", "add", "origin", url,
+                _gitdir=repo)
+
+            git("fetch", "--depth", "1", "--quiet", "origin", commit,
+                _gitdir=repo)
+
+            git("worktree", "add", "--quiet", output, commit,
+                _gitdir=repo)
+
+            # this will turn the worktree into a normal
+            # directory without any reference to git
+            os.unlink(f"{output}/.git")
+
```

Comment:
This does not cache the result anywhere, does it?

Reply:
No, since it is fetching a single commit and that should not be that big I wanted to have this simple for now and now deal with caching.